rodeo/src/rodeo/09-case-classes.scala

```diff
@@ -1,8 +1,112 @@
 package rodeo
 
+import rodeo.Functions.Exercise
+import zio.test.assertTrue
+
+import scala.collection.immutable.List
+
 /**
  *
  */
 object CaseClasses {
+  /*
+      Scala Case Classes are good for modeling immutable data. To define a Case Class we need to specify a constructor
+      parameter list. These parameters are meant to define the most significant properties we want the objects of our
+      class to have. Some of the benefits of using Case Classes are the following.
+
+      1) An accessor method is generated for each parameter in the constructor.
+      2) Case classes are easier to instantiate because  the `new` keyword in unnecessary.
+      3) Case classes allows us to use pattern matching.
+
+  */
```

Comment:
Idea for an exercise would be to have `class Persona(name: String, age: Int)` (a plain class, not a case class), but make changes on it to behave like a case class:

1) Make `name` and `age` accessible outside the instance.
2) Add on companion object an `Persona.apply` constructor.
3) Add on companion object an `Persona.unapply` extractor.
    https://docs.scala-lang.org/tour/extractor-objects.html
4) Add a `copy` method on each instance.

Reply:
I tried to divide this exercise into 4 different exercises. 

I still don't understand why an extractor is fundamental for pattern matching. 
 
Why is it important to be able to recover the parameters that where used to instantiate a class in order to decide which type of input i'm dealing with?

How would you define  pattern matching in the most general case?

Thanks !!!

Reply:
> Why is it important to be able to recover the parameters that where used to instantiate a class in order to decide which type of input i'm dealing with?

Because the instance of the case class you might want to pattern match on it to extract a property from it and perhaps and filter on it, for example:

```scala
// Somewhere far away in the code
sealed trait Shapes
case class Circle(radius: Double) extends Shape
case class Square(side: Double) extends Shape


// In your geometry module
def circlePrinter(circle: Circle) = 
  circle match {
    case Circle(radius) if radius > 10 => println(s"Big Circle of radius ${radius}!")
    case Circle(_) => println("Small Circle!")
    case _ => println("Not a Circle!")
    }
  ```